streams/src/main/java/org/apache/kafka/streams/processor/internals/Task.java

```diff
@@ -26,26 +26,65 @@
 import java.util.Set;
 
 public interface Task {
+    enum State {
+        CREATED, RESTORING, RUNNING, SUSPENDED, CLOSED;
+
+
+        static void validateTransition(final State oldState, final State newState) {
+            if (oldState == CREATED && newState == RESTORING) {
+                return;
+            } else if (oldState == RESTORING && (newState == RESTORING || newState == RUNNING || newState == SUSPENDED)) {
+                return;
+            } else if (oldState == RUNNING && (newState == RESTORING || newState == RUNNING || newState == SUSPENDED)) {
+                return;
+            } else if (oldState == SUSPENDED && (newState == RESTORING || newState == RUNNING || newState == SUSPENDED || newState == CLOSED)) {
+                return;
+            } else {
+                throw new IllegalStateException("Invalid transition from " + oldState + " to " + newState);
+            }
+        }
+    }
+
+    State state();
+
+    void transitionTo(State newState);
+
+    void initializeIfNeeded();
+
+    void startRunning();
 
     void initializeMetadata();
 
     /**
-     * Initialize the task and return {@code true} if the task is ready to run, i.e, it has no state stores
-     * @return true if this task has no state stores that may need restoring.
+     * Initialize the task's stores
      * @throws IllegalStateException If store gets registered after initialized is already finished
      * @throws StreamsException if the store's change log does not contain the partition
      */
-    boolean initializeStateStores();
+    void initializeStateStores();
```

Comment:
I think we can remove initializeStateStores / initializeTopology / initializeMetadata with transitionTo (the initializeIfNeeded can also be private):

1) for for both active and standby transitionTo(Created -> Restoring) always register stores; for active task we also initialize metadata.
2) for active TransitionTo(Restoring -> Running) we initialize the topology.
3) for standby there's no Transition to Running / Suspended, if it ever gets call we throw IllegalState.

Reply:
Done!